Add run script and config template

mary-cleaton opened the pull request:
## Description

Add run script and config template. Minor changes done to run script to allow it to work on Spark 3.5.1 and with S3 buckets instead of HDFS, namely:

- Added `if __name__ = "__main__":` idiom to end of `main.py` so it can be run directly. Thus, `run.py` (which called it) is no longer needed for encapsulation purposes. This also ensures everything is within the `scalelink` folder.
- Added import to the `__init__.py` file within the `scalelink` folder. This allows the `run_scalelink` function in main.py to be imported directly in a script using `from scalelink import run_scalelink` once the code is packaged.
- Switched HDFS code checkpoint clean-up using `subprocess` to a checkpoint clean-up method for s3 bucket files using `raz_client` and `boto3`.
- Updated the required packages list, config template and `read_configs` function to facilitate this new checkpoint clean-up method.
- Updated `create_spark_session` so it works with Spark 3.5.1. Also updated its test.
- Added a configs template and updated it so it will work with the updated `main.py`.
- Updated `test_get_input_variables` to reflect changes in configs.

As this is a main script, there are no unit tests associated with it. However, I have tested that it runs using the synthetic test datasets from the Scalelink R package. **Whoever picks this up for review, please contact me so that you can carry out the same checks.**

## Type of change

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

mary-cleaton commented:
@ONSdigital/scalelink-maintainers and @ONSdigital/scalelink-developers 
See comment in description - contact me for files to run this on synthetic data as a test.

Commit:
ci: stop applying check-yaml pre-commit to github actions

Commit:
fix: add new job to github action
This hopefully should get round the 'subprocess.CalledProcessError: Command 'krb5-config --libs gssapi' returned non-zero exit status 127.' that I keep getting for this branch.

Commit:
ci: update install-librkb5 github action job

As we are always using Ubuntu here, removed the if-clause checking if Ubuntu (or MacOS) was being used. This if-clause had a typo in it and as it wasn't needed, removing it was better than trying to fix the typo.

Commit:
Revert "docs: update changelog"

This reverts commit ec603bfb5c87456815404970b12aebc4d350b3b0.

Commit:
fix: correct issues caused by last few commits

Commit:
fix: add new job to github action
Another attempt at installing libkrb5 to get round the 'subprocess.CalledProcessError: Command 'krb5-config --libs gssapi' returned non-zero exit status 127' error encountered when trying to run the pytests as a GitHub Action now that the code contains a dependency for raz_client.

alani-odunlami reviewed (changes requested):
I'm happy with all the changes that have made to the run script and config templete.
Is it possible  to carryout the checks using the synthetic test datasets from the Scalelink R package?

mary-cleaton commented:
> I'm happy with all the changes that have made to the run script and config templete. Is it possible to carryout the checks using the synthetic test datasets from the Scalelink R package?

Have put the relevant files in your Dev workspace. :)

Commit:
feat: add import to permit easy install
Add import line to scalelink folder's __init__.py so the run_scalelink function can easily be imported from scalelink once it's a package.

Commit:
ci: update required packages

mary-cleaton commented:
Two minor updates made since last QA.

Commit:
ci: remove import line as it seems to have broken things

mary-cleaton commented:
After a lot of back-and-forth, I think the reason the most recent commits have had failing GitHub Actions for running the pytests is that Python 3.9 has been discontinued whilst this branch was waiting for QA.

I'm going to revert this to draft and open up the branch that bumps the Python versions. Once that's updated, this branch can be rebased to it and then we can try again to get the GitHub Actions to pass.

Commit:
docs: undo commenting out of pyspark in package requirements

Commit:
docs: pin pyspark version to 3.5.1

mary-cleaton commented:
Rebased this branch, after merging the branch containing the removal of Python 3.8 and 3.9 and the addition of Python 3.11 to `develop`.

Turned out that this was not the issue causing the `test_calculate_sorensen_dice` unit test to fail. Rather, it was that PySpark was not pinned to 3.5.1 in this branch and it appears PySpark 4.0.0 handles something (possibly NGram) differently. I've raised an issue to fix this in the longer term (see #77) but for this branch I've just pinned the PySpark to 3.5.1. I think this is the better solution, as we don't have capability to do full PySpark 4.0.0 tests locally currently.

danasebo commented:
I think this PR has been superseded by [PR#65](https://github.com/ONSdigital/scalelink/pull/65)

mary-cleaton commented:
> I think this PR has been superseded by [PR#65](https://github.com/ONSdigital/scalelink/pull/65)

Oh bother, you're right. I forgot that the branch in PR #65 was a child of this branch and that this branch should have been merged first. I'll close this one as it's been superseded.
